### PR TITLE
Fix parenthesis simplification around binary expressions

### PR DESCRIPTION
**Pull request by DustinCampbell**
Fixes https://github.com/dotnet/roslyn/issues/11958

This change adds a couple of semantic checks when determining whether parentheses can be removed from around a addition or multiplication expression contained within another addition or multiplication expression. For example, `a + (b + c)` or `a * (b * c)`. Essentially, we're now a bit more conservative and will only remove parentheses if the following semantic rules hold true:

1. The inner binary expression must not result in a user-defined operator being called.
2. The type and converted type of binary expression must match.

This ensures that we don't incorrectly remove parentheses in several cases, such as the following string concatenation example.

```C#
var x = "Hello, world: " + (19 + 23);
```

cc @dotnet/roslyn-ide

### CODE REVIEW

**Comment by DustinCampbell** [comment]
@Pilchie? @CyrusNajmabadi? @roslyn-ide?

**Comment by CyrusNajmabadi** [comment]
Looking.

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 1]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
+class C
+{
+    int Add(int a, short b, short c) => a + b + c;
+}
 </code>
 
             Await TestAsync(input, expected)
         End Function
+
+#End Region
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: why are there so many changes in this file?

**Comment by CyrusNajmabadi** [comment]
Code changes look good.  Don't understand the test changes though...

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb (hunk 1 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: The tests in this file had been organized at one point, but recently added tests weren't added with the same organization. I organized them.

Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: IOW, I moved a bunch of tests around.

File: src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs [hunk 2]
```diff
@@ -4157,5 +4157,31 @@ public class KVP<T1, T2>
 
             await TestAsync(code, expected, index: 0);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTemporary)]
+        [WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")]
+        public async Task EnsureParenthesesInStringConcatenation()
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 3]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
 </code>
 
             Await TestAsync(input, expected)
         End Function
 
-        <WorkItem(4490, "https://github.com/dotnet/roslyn/issues/4490")>
+        <WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")>
         <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
-        Public Async Function TestVisualBasic_DontRemoveParenthesesAroundXmlElementWhenPreviousTokenIsGreaterThan() As Task
+        Public Async Function TestCSharp_SimplifyIfBinaryExpressionTypeIsImplicitNumericConversion() As Task
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 4]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
 </code>
 
             Await TestAsync(input, expected)
         End Function
 
-        <WorkItem(4490, "https://github.com/dotnet/roslyn/issues/4490")>
+        <WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")>
         <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
-        Public Async Function TestVisualBasic_DontRemoveParenthesesAroundXmlElementWhenPreviousTokenIsLessThan() As Task
+        Public Async Function TestCSharp_SimplifyIfBinaryExpressionTypeIsIdentityConversion() As Task
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 5]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
 </code>
 
             Await TestAsync(input, expected)
         End Function
 
-        <WorkItem(4490, "https://github.com/dotnet/roslyn/issues/4490")>
+        <WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")>
         <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
-        Public Async Function TestVisualBasic_DontRemoveParenthesesAroundEmptyXmlElementWhenPreviousTokenIsGreaterThan() As Task
+        Public Async Function TestCSharp_DontSimplifyIfOperatorOverloadsWouldNoLongerByCalled() As Task
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 6]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
 </code>
 
             Await TestAsync(input, expected)
         End Function
 
-        <WorkItem(4490, "https://github.com/dotnet/roslyn/issues/4490")>
+        <WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")>
         <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
-        Public Async Function TestVisualBasic_DontRemoveParenthesesAroundEmptyXmlElementWhenPreviousTokenIsLessThan() As Task
+        Public Async Function TestCSharp_DontSimplifyIfItWouldChangeStringConcatenation() As Task
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 7]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
 </code>
 
             Await TestAsync(input, expected)
         End Function
 
-        <WorkItem(4490, "https://github.com/dotnet/roslyn/issues/4490")>
+        <WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")>
         <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
-        Public Async Function TestVisualBasic_RemoveParenthesesAroundXmlElementInInvocation() As Task
+        Public Async Function TestCSharp_SimplifyInStringConcatenationIfItWouldNotChangeMeaning2() As Task
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb [hunk 8]
```diff
@@ -1042,346 +1297,279 @@ using System;
...
 </code>
 
             Await TestAsync(input, expected)
         End Function
 
-        <WorkItem(4490, "https://github.com/dotnet/roslyn/issues/4490")>
+        <WorkItem(11958, "https://github.com/dotnet/roslyn/issues/11958")>
         <Fact, Trait(Traits.Feature, Traits.Features.Simplification)>
-        Public Async Function TestVisualBasic_RemoveParenthesesAroundXmlElement() As Task
+        Public Async Function TestCSharp_SimplifyInStringConcatenationIfItWouldNotChangeMeaning1() As Task
```
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: Here's a new test

File: src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb (hunk 1 above)
Comment (DustinCampbell) [inline src/EditorFeatures/Test2/Simplification/ParenthesisSimplificationTests.vb]: I marked the newly added tests. Sorry about the diff.

**Comment by CyrusNajmabadi** [comment]
👍

**Comment by DustinCampbell** [comment]
Looking forward to getting some check marks so I can merge. :smile:

**Comment by DustinCampbell** [comment]
I have no idea what we're supposed to target any more. @jasonmalinowski sent mail about branches and then ret-con'd it, but it wasn't clear how much was ret-con'd. So, I don't know what to target.

**Comment by jasonmalinowski** [comment]
master is safe to target, and is probably the right place to go. The only thing that was retconned was dev15.1.x.

**Comment by DustinCampbell** [comment]
Well then, yes, it should be against master 😊

**Comment by DustinCampbell** [comment]
retest this please

**Pull request merged**